README.md

```diff
@@ -1,24 +1,211 @@
+Token.
+
+```html
+<div>Loading...</div>
+```
+
+* Edit the root route in `routes.rb` to point to this loading page as well.
+
+```rb
+Rails.application.routes.draw do
+  resources :widgets
+  root :to => 'application#index'
+  get '/home', to: 'home#index', as :home
+  mount ShopifyApp::Engine, at: '/'
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+end
+```
+
+### Fetching a session token
+
+The preferred method of fetching an authenticated JWT session token is by using
+the `app-bridge-utils` library. It provides the helper method `getSessionToken`
+that fetches a session token for you.
+
+1. Ensure your app includes `app-bridge-utils` with a release version of at
+least [`1.22.0`][6]:
+
+```console
+$ yarn add @shopify/app-bridge-utils@1.22.0
+```
+
+2. Import the `app-bridge-utils` library method `getSessionToken` in
+`apps/javascript/shopify_app/shopify_app.js`:
+
+```js
+import { getSessionToken } from "@shopify/app-bridge-utils";
+```
+
+* `getSessionToken()` takes as argument an instance of app-bridge. If you
+bootstrapped your application using `shopify_app`, one instance can be found
+in `app/javascript/shopify_app/shopify_app.js`. It returns
+an authenticated JWT token.
+* This method does not auto-refresh tokens, so subsequent calls to refresh the
+token will be needed.
+
+3. Use `getSessionToken()` to fetch a JWT token. After the initial successful
+fetch, use Turbolinks to redirect your application to its protected home page
+resource, away from the loading page. This can be done in
+`app/javascript/shopify_app/shopify_app.js`:
+
+```js
+import { getSessionToken } from "@shopify/app-bridge-utils";
+
+document.addEventListener("DOMContentLoaded", async () => {
+  var data = document.getElementById("shopify-app-init").dataset;
+  var AppBridge = window["app-bridge"];
+  var createApp = AppBridge.default;
+  window.app = createApp({
+    apiKey: data.apiKey,
+    shopOrigin: data.shopOrigin,
+  });
+
+  ...
+
+  // Wait for the token before trying to load an authenticated page
+  await retrieveToken(app);
+  Turbolinks.visit("/home");
+
+  // Keep requesting the token every 50 seconds 
+  keepRetrievingToken(app);
+});
+
+async function retrieveToken(app) {
+  window.sessionToken = await getSessionToken(app);
+}
+
+function keepRetrievingToken(app) {
+  setInterval(() => {
+    retrieveToken(app);
+  }, 50000);
+}
+
+document.addEventListener("turbolinks:request-start", function (event) {
+  var xhr = event.data.xhr;
+  xhr.setRequestHeader("Authorization", "Bearer " + window.sessionToken);
+});
+
+document.addEventListener("turbolinks:render", function () {
+  $("form, a[data-method=delete]").on("ajax:beforeSend", function (event) {
+    const xhr = event.detail[0];
+    xhr.setRequestHeader("Authorization", "Bearer " + window.sessionToken);
+  });
+});
+```
+
+* Event listeners are added to both `turbolinks:request-start` and
```

Comment:
Maybe add an explanation for why we have `turbolinks:request-start `and `turbolinks:render` as separate event listeners.